Add support for customizable Solr request handlers.

ojlyytinen opened the pull request:
Lets you configure which Solr request handlers are used for selects and updates by specifying select_path and update_path in solr.yml. This requires a patch in RSolr as well, a pull request for that is at rsolr/rsolr#116. Probably best to wait merging this until that's merged just in case something in there needs to be changed.

I've made this backwards compatible so that Active-Fedora doesn't depend on the new version of RSolr. With an older version of RSolr the custom request handlers simply will not work.

awead commented:
@ojlyytinen can you copy in an example of what the new solr.yml is going to look like with custom update and select path? I'm having trouble visualizing this.

ojlyytinen commented:
Solr.yml would look like something like this. 

```ruby
development:
  url: http://localhost:8983/solr/development
  select_path: select_custom
  update_path: update_custom
```

select_path and update_path can be left out in which case it'll use the defaults.

And you'd have these defined in solrconfig.xml

```xml
  <requestHandler name="/select_custom" class="solr.SearchHandler" default="true">
     <lst name="defaults">
       ...
     </lst>
  </requestHandler>
  <requestHandler name="/update_custom" class="solr.UpdateRequestHandler" >
    ...
  </requestHandler>
```

awead commented:
@ojlyytinen Sorry, I forgot about this one...looks okay to me. If you want to rebase it, I can merge it.

ojlyytinen commented:
I think we should get the RSolr patch merged first. This one won't actually do anything until RSolr supports configurable handlers. Is there anything we can do to expedite that?

ojlyytinen commented:
The RSolr patch has been merged and I have also rebased this. There were a few small changes in this PR as a result of changing how RSolr handles request handlers. Functionally it should be identical though. @awead if you have time to do a review of this that would be much appreciated.

awead commented on lib/active_fedora/solr_service.rb:
```diff
@@ -104,7 +105,8 @@ def construct_query_for_rel(field_pairs, join_with = 'AND')
         args = args.merge(q: query, qt: 'standard')
-        result = SolrService.instance.conn.get('select', params: args)
+        instance = SolrService.instance
+        result = instance.conn.get(instance.select_path, params: args)
```
What's the difference between `@select_path` and `instance.select_path`?

awead commented on spec/unit/finder_methods_spec.rb:
```diff
@@ -47,4 +55,34 @@ def initialize
   end
+
+  describe "#find_in_batches" do
+    let(:docs) {
```
Minor quibble: we use do/end for multiline blocks and {} for single-line blocks.

awead commented on spec/unit/finder_methods_spec.rb:
```diff
@@ -47,4 +55,34 @@ def initialize
+    }
+    let(:select_handler) { 'select' }
+    let(:connection) { double('conn') }
+    before {
```
same here.

awead commented on spec/unit/finder_methods_spec.rb:
```diff
@@ -47,4 +55,34 @@ def initialize
+        .and_return('response' => { 'docs' => docs })
+    }
+    it "yields the docs" do
+      expect { |b|
```
I think it's okay in this case, tho.

ojlyytinen commented:
I moved select_path from SolrService.instance to SolrService class level since it wasn't actually ever used in any of the instance methods. I think it's a bit cleaner that way.

I also cleaned up the block formatting issues in specs.

awead commented:
@ojlyytinen :clap:
